### PR TITLE
Fixing error of array sort with compare function (resolves #266) 

### LINKED ISSUE #266: array sort function bug with custom compare function
Array sorting is wrong if compare function returns abs(value) < 1 or abs(value) > 2147483647.
```
package main

import "github.com/robertkrimen/otto"

func main() {
vm := otto.New()
vm.Run(`
	console.log([0.5, 0.8, 0.2].sort(function (a,b) { return a-b; }));
	console.log([2147483647, -1, 1].sort(function (a,b) { return a-b; }));
	console.log([2e10, -2e10, 0].sort(function (a,b) { return a-b; }));
`)
}
```
produces
```
0.5,0.2,0.8
2147483647,-1,1
0,-200000000000,200000000000
```

### PR DESCRIPTION
**Pull request by limexp**
* Test for issue #266 
* New func toIntSign(..) to get value sign
* Updated sortCompare result from int(toInt32(..)) call to toIntSign(...)

### CODE REVIEW

File: value_number.go [hunk 1]
```diff
@@ -322,3 +322,72 @@ func toUint16(value Value) uint16 {
...
+		return 0
+	}
+	if floatValue == 0 {
+		return 0
+	}
+	if floatValue > 0 {
+		return 1
+	} else {
+		return -1
+	}
```
Comment (stevenh) [inline value_number.go:391]:
```suggestion
	} 
	return -1
```

File: value_number.go [hunk 2]
```diff
@@ -322,3 +322,72 @@ func toUint16(value Value) uint16 {
 	}
 	return uint16(remainder)
 }
+
+// Returns sign of a number converted to -1,0,1
```
Comment (stevenh) [inline value_number.go:326]:
```suggestion
// toIntSign returns sign of a number converted to -1, 0 ,1
```

File: value_number.go [hunk 3]
```diff
@@ -322,3 +322,72 @@ func toUint16(value Value) uint16 {
...
+	{
+		switch value := value.value.(type) {
+		case int8:
+			if int8(value) > 0 {
+				return 1
+			} else if int8(value) < 0 {
+				return -1
+			} else {
+				return 0
+			}
```
Comment (stevenh) [inline value_number.go:337]:
```suggestion
			}

			return 0
```

File: value_number.go [hunk 4]
```diff
@@ -322,3 +322,72 @@ func toUint16(value Value) uint16 {
...
+		case int16:
+			if int16(value) > 0 {
+				return 1
+			} else if int16(value) < 0 {
+				return -1
+			} else {
+				return 0
+			}
```
Comment (stevenh) [inline value_number.go:345]:
```suggestion
			} 
				
			return 0
```

File: value_number.go [hunk 5]
```diff
@@ -322,3 +322,72 @@ func toUint16(value Value) uint16 {
...
+		case int32:
+			if int32(value) > 0 {
+				return 1
+			} else if int32(value) < 0 {
+				return -1
+			} else {
+				return 0
+			}
```
Comment (stevenh) [inline value_number.go:353]:
```suggestion
			} 
				
			return 0
```

File: value_number.go [hunk 6]
```diff
@@ -322,3 +322,72 @@ func toUint16(value Value) uint16 {
...
+		case uint8:
+			if uint8(value) > 0 {
+				return 1
+			} else if uint8(value) < 0 {
+				return -1
+			} else {
+				return 0
+			}
```
Comment (stevenh) [inline value_number.go:361]:
```suggestion
			} 
				
			return 0
```

File: value_number.go [hunk 7]
```diff
@@ -322,3 +322,72 @@ func toUint16(value Value) uint16 {
...
+		case uint16:
+			if uint16(value) > 0 {
+				return 1
+			} else if uint16(value) < 0 {
+				return -1
+			} else {
+				return 0
+			}
```
Comment (stevenh) [inline value_number.go:369]:
```suggestion
			} 
				
			return 0
```

File: value_number.go [hunk 8]
```diff
@@ -322,3 +322,72 @@ func toUint16(value Value) uint16 {
...
+		case uint32:
+			if uint32(value) > 0 {
+				return 1
+			} else if uint32(value) < 0 {
+				return -1
+			} else {
+				return 0
+			}
```
Comment (stevenh) [inline value_number.go:377]:
```suggestion
			} 
				
			return 0
```

**Review by stevenh — CHANGES_REQUESTED** [review]

**Comment by limexp** [comment]
Requested changes applied

**Comment by Asday** [comment]
Yo @stevenh you've been going berserk today.  Good work!  You ok?

File: value_number.go [hunk 9]
```diff
@@ -300,3 +300,71 @@ func toUint16(value Value) uint16 {
...
+		case uint32:
+			if uint32(value) > 0 {
+				return 1
+			} else if uint32(value) < 0 {
+				return -1
+			} 
+				
+			return 0
+		}
+	}
```
Comment (stevenh) [inline value_number.go:357]:
```suggestion
	switch value := value.value.(type) {
	case int8:
		if int8(value) > 0 {
			return 1
		} else if int8(value) < 0 {
			return -1
		}
	
		return 0
	case int16:
		if int16(value) > 0 {
			return 1
		} else if int16(value) < 0 {
			return -1
		} 
			
		return 0
	case int32:
		if int32(value) > 0 {
			return 1
		} else if int32(value) < 0 {
			return -1
		} 
			
		return 0
	case uint8:
		if uint8(value) > 0 {
			return 1
		} else if uint8(value) < 0 {
			return -1
		} 
			
		return 0
	case uint16:
		if uint16(value) > 0 {
			return 1
		} else if uint16(value) < 0 {
			return -1
		} 
			
		return 0
	case uint32:
		if uint32(value) > 0 {
			return 1
		} else if uint32(value) < 0 {
			return -1
		} 
			
		return 0
	}
```

File: value_number.go [hunk 10]
```diff
@@ -300,3 +300,71 @@ func toUint16(value Value) uint16 {
...
+	if math.IsNaN(floatValue) || math.IsInf(floatValue, 0) {
+		return 0
+	}
+	if floatValue == 0 {
+		return 0
+	}
+	if floatValue > 0 {
+		return 1
+	} 
+	return -1
```
Comment (stevenh) [inline value_number.go:368]:
```suggestion
	switch {
	case math.IsNaN(floatValue), math.IsInf(floatValue, 0):
		return 0
	case floatValue == 0:
		return 0
	case floatValue > 0:
		return 1
	default:
		return -1
	}
```

**Review by stevenh — CHANGES_REQUESTED** [review]
Sorry I missed the over indented `switch` and `if {} else if {}` to `switch {}` clean up opportunity.

**Comment by stevenh** [comment]
> Yo @stevenh you've been going berserk today. Good work! You ok?

LOL, I noticed I'd not been getting emails for these issues / PR so thought I'd have a look in some spare time to try and close some long overdue issues out ;-)

**Comment by limexp** [comment]
> If you could also rebase to master so it picks up the github action tests, that would be great.

Rebased. Thank you for reviews.

**Review by stevenh — APPROVED** [review]

**Pull request merged**